<a href="https://colab.research.google.com/github/Khomkrit007/bsc_dpdm24/blob/main/Model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Import Data

In [1]:
# prompt: เชื่อมต่อไปยัง drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [82]:
path_to_train ='/content/drive/MyDrive/มหาลัยขอนแก่น/3-2/Data prepare & Data Mining/สร้าง model/train_set.csv'
path_to_test ='/content/drive/MyDrive/มหาลัยขอนแก่น/3-2/Data prepare & Data Mining/สร้าง model/X_test.csv'
path_to_y_pred ='/content/drive/MyDrive/มหาลัยขอนแก่น/3-2/Data prepare & Data Mining/สร้าง model/y_pred_baseline.csv'

In [3]:
import pandas as pd

In [164]:
train_df = pd.read_csv(path_to_train)
test_df = pd.read_csv(path_to_test)
y_pred_df = pd.read_csv(path_to_y_pred)

In [160]:
train_df

id number    Xn  Hb(Gm%)  Hct(%)  MCV(fl)  A2 (2-3%)  HbF (2%)  \
0       67d169  25.0     11.1    34.0     78.0        3.2       0.0   
1       67l168  32.0     13.9    42.0     86.0        3.7       0.3   
2       67i267  43.0     10.1    31.0     78.0        2.6       0.3   
3       67e248  33.0     12.9    38.0     76.0        3.0       0.2   
4       67h277  23.0     12.9    40.0     78.0        3.1       0.0   
...        ...   ...      ...     ...      ...        ...       ...   
4317    66z431  63.0     10.1    31.0     83.0        2.8       1.8   
4318    66z436  19.0     11.1    34.0     79.0        2.6       0.6   
4319    66z439  80.0     10.9    33.9     79.0        3.0       0.2   
4320    66z445  27.0     13.2    41.4     87.0        2.9       0.3   
4321     66z66  29.0      7.2    22.9     78.0        2.4       0.0   

      HbA (95-97%)       RBC        MCH       MCHC  MCV-RBC-3*HGB  \
0             55.9  4.358974  25.464706  32.647059      40.341026   
1             54.8  5.197674  28.859060  33.557047      35.802326   
2             56.2  3.974359  25.412903  32.580645      43.725641   
3             61.2  5.000000  25.800000  33.947368      32.300000   
4             55.6  5.128205  25.155000  32.250000      34.171795   
...            ...       ...        ...        ...            ...   
4317          84.2  3.734940  27.041935  32.580645      48.965060   
4318          78.0  4.303797  25.791176  32.647059      41.396203   
4319          87.2  4.291139  25.401180  32.153392      42.008861   
4320          87.6  4.758621  27.739130  31.884058      42.641379   
4321          97.6  3.100602  24.524017  31.441048      53.464103   

      MCV-RBC-5*HGB  |80-MCV|  |27-MCH|  |80-MCV|*|27-MCH|         Thal_Group  
0         18.141026       2.0  1.535294           3.070588  Abnormal_variants  
1          5.802326       6.0  1.859060          11.154362  Abnormal_variants  
2         23.525641       2.0  1.587097           3.174194  Abnormal_variants  
3          6.500000       4.0  1.200000           4.800000  Abnormal_variants  
4          8.371795       2.0  1.845000           3.690000  Abnormal_variants  
...             ...       ...       ...                ...                ...  
4317      28.765060       3.0  0.041935           0.125806             Normal  
4318      19.196203       1.0  1.208824           1.208824             Normal  
4319      20.208861       1.0  1.598820           1.598820             Normal  
4320      16.241379       7.0  0.739130           5.173913             Normal  
4321      35.679858       2.0  2.475983           4.951965             Normal  

[4322 rows x 17 columns]

In [52]:
test_df

id number    Xn  Hb(Gm%)  Hct(%)  MCV(fl)  A2 (2-3%)  HbF (2%)  \
0       66z133  38.0     12.4   37.00     78.0        2.7       1.3   
1       66z297  28.0     12.8   39.40     82.0        2.7       0.0   
2       66u491  78.0     10.9   33.90     79.0        3.1       0.0   
3       66x072  65.0     10.9   33.90     79.0        3.0       0.2   
4       67m031  17.0     11.5   36.00     78.0        2.8       0.8   
...        ...   ...      ...     ...      ...        ...       ...   
1992    67j317   NaN      8.4   26.00     59.0        2.6       0.7   
1993    66a385   NaN     10.1   31.00     78.0        3.3       0.0   
1994    67g355  53.0     10.4   34.00     77.0        3.0       0.2   
1995    66s121  54.0     10.1   31.00     78.0        2.4       0.3   
1996    67h101  64.0     11.9   37.85     65.5       21.7       1.1   

      HbA (95-97%)       RBC        MCH       MCHC  MCV-RBC-3*HGB  \
0             75.4  4.743590  26.140541  33.513514      36.056410   
1             97.3  4.804878  26.639594  32.487310      38.795122   
2             96.9  4.291139  25.401180  32.153392      42.008861   
3             87.9  4.291139  25.401180  32.153392      42.008861   
4             79.6  4.615385  24.916667  31.944444      38.884615   
...            ...       ...        ...        ...            ...   
1992           0.3  4.406780  19.061538  32.307692      29.393220   
1993          55.0  3.974359  25.412903  32.580645      43.725641   
1994          49.6  4.415584  23.552941  30.588235      41.384416   
1995          42.0  3.974359  25.412903  32.580645      43.725641   
1996          66.1  5.778626  20.593131  31.439894      24.021374   

      MCV-RBC-5*HGB  |80-MCV|  |27-MCH|  |80-MCV|*|27-MCH|  
0         11.256410       2.0  0.859459           1.718919  
1         13.195122       2.0  0.360406           0.720812  
2         20.208861       1.0  1.598820           1.598820  
3         20.208861       1.0  1.598820           1.598820  
4         15.884615       2.0  2.083333           4.166667  
...             ...       ...       ...                ...  
1992      12.593220      21.0  7.938462         166.707692  
1993      23.525641       2.0  1.587097           3.174194  
1994      20.584416       3.0  3.447059          10.341176  
1995      23.525641       2.0  1.587097           3.174194  
1996       0.221374      14.5  6.406869          92.899604  

[1997 rows x 16 columns]

In [7]:
y_pred_df

Thal_Group
0        Normal
1        Normal
2        Normal
3        Normal
4        Normal
...         ...
1992     Normal
1993     Normal
1994     Normal
1995     Normal
1996     Normal

[1997 rows x 1 columns]

# Cleaning Data

In [40]:
train_df.isnull().sum()

id number              0
Xn                   435
Hb(Gm%)                7
Hct(%)                 7
MCV(fl)                7
A2 (2-3%)            133
HbF (2%)             116
HbA (95-97%)         116
RBC                    7
MCH                    7
MCHC                   7
MCV-RBC-3*HGB          7
MCV-RBC-5*HGB          7
|80-MCV|               7
|27-MCH|               7
|80-MCV|*|27-MCH|      7
Thal_Group             0
dtype: int64

In [41]:
test_df.isnull().sum()

id number              0
Xn                   187
Hb(Gm%)                2
Hct(%)                 2
MCV(fl)                1
A2 (2-3%)             62
HbF (2%)              59
HbA (95-97%)          59
RBC                    2
MCH                    2
MCHC                   2
MCV-RBC-3*HGB          2
MCV-RBC-5*HGB          2
|80-MCV|               1
|27-MCH|               2
|80-MCV|*|27-MCH|      2
dtype: int64

In [165]:
# จัดการค่า Missing โดยแทนที่ด้วยค่าเฉลี่ยของแต่ละคอลัมน์สำหรับคอลัมน์ที่เป็นตัวเลขเท่านั้น
numeric_columns = train_df.select_dtypes(include=['number']).columns
train_df[numeric_columns] = train_df[numeric_columns].fillna(train_df[numeric_columns].mean())
test_df[numeric_columns] = test_df[numeric_columns].fillna(test_df[numeric_columns].mean())

In [155]:
train_df.isnull().sum()

Xn                   0
Hb(Gm%)              0
Hct(%)               0
MCV(fl)              0
A2 (2-3%)            0
HbF (2%)             0
HbA (95-97%)         0
RBC                  0
MCH                  0
MCHC                 0
MCV-RBC-3*HGB        0
MCV-RBC-5*HGB        0
|80-MCV|             0
|27-MCH|             0
|80-MCV|*|27-MCH|    0
Thal_Group           0
dtype: int64

In [156]:
test_df.isnull().sum()

Xn                   0
Hb(Gm%)              0
Hct(%)               0
MCV(fl)              0
A2 (2-3%)            0
HbF (2%)             0
HbA (95-97%)         0
RBC                  0
MCH                  0
MCHC                 0
MCV-RBC-3*HGB        0
MCV-RBC-5*HGB        0
|80-MCV|             0
|27-MCH|             0
|80-MCV|*|27-MCH|    0
dtype: int64

In [162]:
# Display the normalized dataframes
train_df

id number    Xn  Hb(Gm%)  Hct(%)  MCV(fl)  A2 (2-3%)  HbF (2%)  \
0       67d169  25.0     11.1    34.0     78.0        3.2       0.0   
1       67l168  32.0     13.9    42.0     86.0        3.7       0.3   
2       67i267  43.0     10.1    31.0     78.0        2.6       0.3   
3       67e248  33.0     12.9    38.0     76.0        3.0       0.2   
4       67h277  23.0     12.9    40.0     78.0        3.1       0.0   
...        ...   ...      ...     ...      ...        ...       ...   
4317    66z431  63.0     10.1    31.0     83.0        2.8       1.8   
4318    66z436  19.0     11.1    34.0     79.0        2.6       0.6   
4319    66z439  80.0     10.9    33.9     79.0        3.0       0.2   
4320    66z445  27.0     13.2    41.4     87.0        2.9       0.3   
4321     66z66  29.0      7.2    22.9     78.0        2.4       0.0   

      HbA (95-97%)       RBC        MCH       MCHC  MCV-RBC-3*HGB  \
0             55.9  4.358974  25.464706  32.647059      40.341026   
1             54.8  5.197674  28.859060  33.557047      35.802326   
2             56.2  3.974359  25.412903  32.580645      43.725641   
3             61.2  5.000000  25.800000  33.947368      32.300000   
4             55.6  5.128205  25.155000  32.250000      34.171795   
...            ...       ...        ...        ...            ...   
4317          84.2  3.734940  27.041935  32.580645      48.965060   
4318          78.0  4.303797  25.791176  32.647059      41.396203   
4319          87.2  4.291139  25.401180  32.153392      42.008861   
4320          87.6  4.758621  27.739130  31.884058      42.641379   
4321          97.6  3.100602  24.524017  31.441048      53.464103   

      MCV-RBC-5*HGB  |80-MCV|  |27-MCH|  |80-MCV|*|27-MCH|         Thal_Group  
0         18.141026       2.0  1.535294           3.070588  Abnormal_variants  
1          5.802326       6.0  1.859060          11.154362  Abnormal_variants  
2         23.525641       2.0  1.587097           3.174194  Abnormal_variants  
3          6.500000       4.0  1.200000           4.800000  Abnormal_variants  
4          8.371795       2.0  1.845000           3.690000  Abnormal_variants  
...             ...       ...       ...                ...                ...  
4317      28.765060       3.0  0.041935           0.125806             Normal  
4318      19.196203       1.0  1.208824           1.208824             Normal  
4319      20.208861       1.0  1.598820           1.598820             Normal  
4320      16.241379       7.0  0.739130           5.173913             Normal  
4321      35.679858       2.0  2.475983           4.951965             Normal  

[4322 rows x 17 columns]

In [163]:
test_df

id number         Xn  Hb(Gm%)  Hct(%)  MCV(fl)  A2 (2-3%)  HbF (2%)  \
0       66z133  38.000000     12.4   37.00     78.0        2.7       1.3   
1       66z297  28.000000     12.8   39.40     82.0        2.7       0.0   
2       66u491  78.000000     10.9   33.90     79.0        3.1       0.0   
3       66x072  65.000000     10.9   33.90     79.0        3.0       0.2   
4       67m031  17.000000     11.5   36.00     78.0        2.8       0.8   
...        ...        ...      ...     ...      ...        ...       ...   
1992    67j317  40.675138      8.4   26.00     59.0        2.6       0.7   
1993    66a385  40.675138     10.1   31.00     78.0        3.3       0.0   
1994    67g355  53.000000     10.4   34.00     77.0        3.0       0.2   
1995    66s121  54.000000     10.1   31.00     78.0        2.4       0.3   
1996    67h101  64.000000     11.9   37.85     65.5       21.7       1.1   

      HbA (95-97%)       RBC        MCH       MCHC  MCV-RBC-3*HGB  \
0             75.4  4.743590  26.140541  33.513514      36.056410   
1             97.3  4.804878  26.639594  32.487310      38.795122   
2             96.9  4.291139  25.401180  32.153392      42.008861   
3             87.9  4.291139  25.401180  32.153392      42.008861   
4             79.6  4.615385  24.916667  31.944444      38.884615   
...            ...       ...        ...        ...            ...   
1992           0.3  4.406780  19.061538  32.307692      29.393220   
1993          55.0  3.974359  25.412903  32.580645      43.725641   
1994          49.6  4.415584  23.552941  30.588235      41.384416   
1995          42.0  3.974359  25.412903  32.580645      43.725641   
1996          66.1  5.778626  20.593131  31.439894      24.021374   

      MCV-RBC-5*HGB  |80-MCV|  |27-MCH|  |80-MCV|*|27-MCH|  
0         11.256410       2.0  0.859459           1.718919  
1         13.195122       2.0  0.360406           0.720812  
2         20.208861       1.0  1.598820           1.598820  
3         20.208861       1.0  1.598820           1.598820  
4         15.884615       2.0  2.083333           4.166667  
...             ...       ...       ...                ...  
1992      12.593220      21.0  7.938462         166.707692  
1993      23.525641       2.0  1.587097           3.174194  
1994      20.584416       3.0  3.447059          10.341176  
1995      23.525641       2.0  1.587097           3.174194  
1996       0.221374      14.5  6.406869          92.899604  

[1997 rows x 16 columns]

In [170]:
import pandas as pd
import numpy as np
from sklearn.model_selection import cross_validate, StratifiedKFold
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, make_scorer
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split, cross_val_score

In [178]:
# ลบคอลัมน์ที่ไม่ใช่ feature
train_df.drop(columns=['id number'], inplace=True, errors='ignore')
test_df.drop(columns=['id number'], inplace=True, errors='ignore')

In [179]:
X = train_df.drop('Thal_Group', axis=1)
y = train_df['Thal_Group']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [180]:
# แบ่งข้อมูลเป็นชุดฝึกและชุดทดสอบ
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [181]:
# กำหนด scoring metrics สำหรับการประเมิน
scoring = {
    'accuracy': 'accuracy',
    'precision_macro': 'precision_macro',
    'recall_macro': 'recall_macro',
    'f1_macro': 'f1_macro'
}

In [182]:
# สร้าง list ของโมเดลพร้อม Hyperparameter Tuning
models = {
    "Decision Tree": DecisionTreeClassifier(max_depth=10, min_samples_split=5, min_samples_leaf=2, random_state=42),
    "Random Forest": RandomForestClassifier(n_estimators=50, max_depth=10, min_samples_split=5, random_state=42),
    "Logistic Regression": LogisticRegression(C=0.1, solver='liblinear', random_state=42),
    "SVC": SVC(C=0.1, kernel='rbf', random_state=42),
    "KNN": KNeighborsClassifier(n_neighbors=5)
}


In [183]:
# สร้าง DataFrame เปล่าสำหรับเก็บผลลัพธ์
results = pd.DataFrame(columns=['Model', 'Accuracy', 'Precision', 'Recall', 'F1-score'])

In [184]:
# ทำ Cross-Validation และเก็บผลลัพธ์
# กำหนดใช้ StratifiedKFold เพื่อรักษาสัดส่วนของแต่ละคลาสในการแบ่งข้อมูล
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

In [185]:
# ปรับขนาดข้อมูล
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [186]:
for name, model in models.items():
    print(f"ประเมินโมเดล {name}...")

    # ทำ cross-validation
    cv_results = cross_validate(
        model, X_scaled, y,
        cv=cv,
        scoring=scoring,
        return_train_score=False
    )

    # คำนวณค่าเฉลี่ยของผลลัพธ์จาก cross-validation
    accuracy = np.mean(cv_results['test_accuracy'])
    precision = np.mean(cv_results['test_precision_macro'])
    recall = np.mean(cv_results['test_recall_macro'])
    f1 = np.mean(cv_results['test_f1_macro'])

    # เพิ่มผลลัพธ์ลงใน DataFrame
    results = pd.concat([results, pd.DataFrame({
        'Model': [name],
        'Accuracy': [accuracy],
        'Precision': [precision],
        'Recall': [recall],
        'F1-score': [f1]
    })], ignore_index=True)

ประเมินโมเดล Decision Tree...


/usr/local/lib/python3.11/dist-packages/sklearn/model_selection/_split.py:805: UserWarning: The least populated class in y has only 3 members, which is less than n_splits=5.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predic

ประเมินโมเดล Random Forest...


/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


ประเมินโมเดล Logistic Regression...


/usr/local/lib/python3.11/dist-packages/sklearn/model_selection/_split.py:805: UserWarning: The least populated class in y has only 3 members, which is less than n_splits=5.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predic

ประเมินโมเดล SVC...


/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/m

ประเมินโมเดล KNN...


/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


In [187]:
print("\nผลการประเมินโมเดลด้วย 5-Fold Cross-Validation:")
pd.set_option('display.precision', 4)
print(results)


ผลการประเมินโมเดลด้วย 5-Fold Cross-Validation:
                 Model  Accuracy  Precision  Recall  F1-score
0        Decision Tree    0.9107     0.7811  0.7806    0.7758
1        Random Forest    0.9301     0.8526  0.8026    0.8168
2  Logistic Regression    0.7788     0.4698  0.4739    0.4538
3                  SVC    0.7744     0.5772  0.4737    0.4610
4                  KNN    0.8607     0.7173  0.6335    0.6519


In [188]:
# เลือกโมเดลที่ดีที่สุด (โดยพิจารณาจาก F1-score)
best_model_idx = results['F1-score'].idxmax()
best_model_name = results.loc[best_model_idx, 'Model']
print(f"\nโมเดลที่ดีที่สุดคือ: {best_model_name}")


โมเดลที่ดีที่สุดคือ: Random Forest


โมเดลที่ดีที่สุด คือ RandomForest มีค่า F1-score สูงที่สุด 0.8168

# ทำนาย Thal_Group จาก ข้อมูลใน X_test

In [144]:
test_df

Xn   Hb(Gm%)    Hct(%)   MCV(fl)  A2 (2-3%)  HbF (2%)  \
0     0.391753  0.600000  0.674089  0.530303      0.027  0.013105   
1     0.288660  0.622222  0.722672  0.590909      0.027  0.000000   
2     0.804124  0.516667  0.611336  0.545455      0.031  0.000000   
3     0.670103  0.516667  0.611336  0.545455      0.030  0.002016   
4     0.175258  0.550000  0.653846  0.530303      0.028  0.008065   
...        ...       ...       ...       ...        ...       ...   
1992  0.419331  0.377778  0.451417  0.242424      0.026  0.007056   
1993  0.419331  0.472222  0.552632  0.530303      0.033  0.000000   
1994  0.546392  0.488889  0.613360  0.515152      0.030  0.002016   
1995  0.556701  0.472222  0.552632  0.530303      0.024  0.003024   
1996  0.659794  0.572222  0.691296  0.340909      0.217  0.011089   

      HbA (95-97%)       RBC       MCH      MCHC  MCV-RBC-3*HGB  \
0         0.634146  0.591835  0.630370  0.823324       0.393617   
1         0.818335  0.600176  0.648411  0.770866       0.428292   
2         0.814971  0.530256  0.603640  0.753796       0.468982   
3         0.739277  0.530256  0.603640  0.753796       0.468982   
4         0.669470  0.574386  0.586125  0.743115       0.429425   
...            ...       ...       ...       ...            ...   
1992      0.002523  0.545995  0.374452  0.761684       0.309253   
1993      0.462574  0.487142  0.604064  0.775637       0.490718   
1994      0.417157  0.547193  0.536824  0.673788       0.461076   
1995      0.353238  0.487142  0.604064  0.775637       0.490718   
1996      0.555929  0.732703  0.429822  0.717323       0.241239   

      MCV-RBC-5*HGB  |80-MCV|  |27-MCH|  |80-MCV|*|27-MCH|  
0          0.327411  0.046512  0.046975           0.003132  
1          0.347505  0.046512  0.019698           0.001313  
2          0.420200  0.023256  0.087385           0.002913  
3          0.420200  0.023256  0.087385           0.002913  
4          0.375381  0.046512  0.113866           0.007591  
...             ...       ...       ...                ...  
1992       0.341267  0.488372  0.433884           0.303718  
1993       0.454577  0.046512  0.086744           0.005783  
1994       0.424092  0.069767  0.188402           0.018840  
1995       0.454577  0.046512  0.086744           0.005783  
1996       0.213038  0.337209  0.350173           0.169250  

[1997 rows x 15 columns]

In [193]:
# เทรนโมเดลที่ดีที่สุดด้วยข้อมูลทั้งหมดและทำนายข้อมูล X_test
best_model = models[best_model_name]
best_model.fit(X_scaled, y)
# ปรับขนาดข้อมูล X_test และทำนาย
X_test_scaled = scaler.transform(X_test)
y_test_pred = best_model.predict(X_test_scaled)


In [194]:
# Convert y_test and y_pred to a consistent data type (e.g., string)
y_test = y_test.astype(str)
y_pred = y_pred.astype(str)

print(classification_report(y_test, y_test_pred)) # Use y_test_pred instead of y_pred

                    precision    recall  f1-score   support

 Abnormal_variants       1.00      1.00      1.00         3
Alpha_thal_related       0.94      0.98      0.96       201
 Beta_thal_related       1.00      1.00      1.00        96
       HbE_related       1.00      1.00      1.00       230
            Normal       0.98      0.96      0.97       335

          accuracy                           0.98       865
         macro avg       0.99      0.99      0.99       865
      weighted avg       0.98      0.98      0.98       865



In [195]:
predictions_df = pd.DataFrame({'Thal_Group_Predicted': y_test_pred})
predictions_df.to_csv('thal_group_predictions.csv', index=False)

print(f"\nการทำนายด้วยโมเดล {best_model_name} สำเร็จแล้ว")
print("ผลการทำนายถูกบันทึกลงในไฟล์ 'thal_group_predictions.csv'")


การทำนายด้วยโมเดล Random Forest สำเร็จแล้ว
ผลการทำนายถูกบันทึกลงในไฟล์ 'thal_group_predictions.csv'
